src/java.base/share/classes/jdk/internal/classfile/impl/verifier/ParserVerifier.java

```diff
@@ -0,0 +1,373 @@
+            }
+            default -> {}
+        }
+    }
+
+    private void verifyAttribute(AttributedElement ae, Attribute<?> a, List<VerifyError> errors) {
+        int payLoad = ((BoundAttribute)a).payloadLen();
+        if (payLoad != switch (a) {
+            case AnnotationDefaultAttribute aa -> valueSize(aa.defaultValue());
+            case BootstrapMethodsAttribute bma -> {
+                int l = 2;
+                for (var bm : bma.bootstrapMethods()) {
+                    l += 4 + 2 * bm.arguments().size();
+                }
+                yield l;
+            }
+            case CharacterRangeTableAttribute cra -> 2 + 14 * cra.characterRangeTable().size();
+            case CodeAttribute ca -> {
+                MethodModel mm = (MethodModel)ae;
+                if (mm.flags().has(AccessFlag.NATIVE) || mm.flags().has(AccessFlag.ABSTRACT)) {
+                    errors.add(new VerifyError("Code attribute in native or abstract %s".formatted(toString(ae))));
+                }
+                if (ca.maxLocals() < Util.maxLocals(mm.flags().flagsMask(), mm.methodTypeSymbol())) {
+                    errors.add(new VerifyError("Arguments can't fit into locals in %s".formatted(toString(ae))));
+                }
+                int l = 12 + ca.codeLength() + 8 * ca.exceptionHandlers().size();
+                for (var caa : ca.attributes()) {
+                    l += 6 + ((BoundAttribute)caa).payloadLen();
+                }
+                yield l;
+            }
+            case CompilationIDAttribute cida -> 2;
+            case ConstantValueAttribute cva -> {
+                ClassDesc type = ((FieldModel)ae).fieldTypeSymbol();
+                ConstantValueEntry cve = cva.constant();
+                if (!switch (TypeKind.from(type)) {
+                    case BooleanType, ByteType, CharType, IntType, ShortType -> cve instanceof IntegerEntry;
+                    case DoubleType -> cve instanceof DoubleEntry;
+                    case FloatType -> cve instanceof FloatEntry;
+                    case LongType -> cve instanceof LongEntry;
+                    case ReferenceType -> type.equals(ConstantDescs.CD_String) && cve instanceof StringEntry;
+                    case VoidType -> false;
+                }) {
+                    errors.add(new VerifyError("Bad constant value type in %s".formatted(toString(ae))));
+                }
+                yield 2;
+            }
+            case DeprecatedAttribute da -> 0;
+            case EnclosingMethodAttribute ema -> 4;
+            case ExceptionsAttribute ea -> 2 + 2 * ea.exceptions().size();
+            case InnerClassesAttribute ica -> {
+                for (var ici : ica.classes()) {
+                    if (ici.outerClass().isPresent() && ici.outerClass().get().equals(ici.innerClass())) {
+                        errors.add(new VerifyError("Class is both outer and inner class in %s".formatted(toString(ae))));
```

Comment:
nit: in `classFileParser` the string includes "class file" after "in". I was just wondering if "class file" is generated by `toString`. I see only `default -> "class " + className();` in the implementation of `toString`. Ignore if the actual message is not important.

Reply:
The `AttributedElement` argument only adds specific info about the enclosing element, so for the class-level attribute errors it falls to the default. Your can se examples of this formatting  in the test:
https://github.com/openjdk/jdk/blob/586fa4b3241a0fc9a2aeddc41d79a74f2259ae52/test/jdk/jdk/classfile/VerifierSelfTest.java#L177